### PR TITLE
Try to bypass DDoS-GUARD

### PR DESCRIPTION
**Pull request by Mystic-Mirage**
Based on https://git.gay/a/ddos-guard-bypass/src/branch/master/index.js

### CODE REVIEW

**Comment by Mystic-Mirage** [comment]
Need to test, because it can degrade performance.

File: ddoser.py [hunk 1]
```diff
@@ -20,12 +20,15 @@
...
 from aiohttp_socks import ProxyConnector
+from cachetools import TTLCache
 
+import ddos_guard
 from commons import config_logger, set_limits, load_proxies, Proxy
 
 STATS = defaultdict(int)
 URL_ERRORS_COUNT = defaultdict(int)
 URL_STATUS_STATS = defaultdict(lambda: defaultdict(int))
+DDOS_GUARD_COOKIE_CACHE = TTLCache(maxsize=100, ttl=60)
```
Comment (Mystic-Mirage) [inline ddoser.py]: Does it make sense? Does it work at all?

Comment (Mystic-Mirage) [reply ddoser.py]: I mean due to parallelism

File: ddos_guard.py [hunk 2]
```diff
@@ -0,0 +1,73 @@
...
+    headers_2 = {
+        "Accept": "*/*",
+        "Accept-Language": "en-US,en;q=0.5",
+        "Accept-Encoding": "gzip, deflate",
+        "Referer": url,
+        "Sec-Fetch-Dest": "script",
+        "Sec-Fetch-Mode": "no-cors",
+        "Sec-Fetch-Site": "cross-site"
+    }
+    async with session.get("https://check.ddos-guard.net/check.js", cookies=cookies, headers=headers_2) as response_2:
```
Comment (e27182) [inline ddos_guard.py]: I suggest `cookies=None` should be here, as it is 1st request to this domain.

File: ddos_guard.py [hunk 3]
```diff
@@ -0,0 +1,73 @@
...
+        text = await response_2.text()
+
+    match = re.search(r"'(?P<path>/.well-known/ddos-guard/id/\S+)'", text)
```
Comment (e27182) [inline ddos_guard.py]: There is a `etag` header in response, it contains id we need. Also this id available as `__ddg2` cookie value.

Comment (e27182) [reply ddos_guard.py]: Also, below we can live without "path" from regexp as it is standard one: "/.well-known/ddos-guard/id/{__ddg2}"

File: ddos_guard.py [hunk 4]
```diff
@@ -0,0 +1,73 @@
...
+
+    headers_3 = {
+        "Accept": "image/webp,*/*",
+        "Accept-Language": "en-US,en;q=0.5",
+        "Accept-Encoding": "gzip, deflate",
+        "Cache-Control": "no-cache",
+        "Referer": url,
+        "Sec-Fetch-Dest": "script",
+        "Sec-Fetch-Mode": "no-cors",
+        "Sec-Fetch-Site": "cross-site"
```
Comment (e27182) [inline ddos_guard.py]: "Sec-Fetch-Site": "cross-site" -> "Sec-Fetch-Site": "same-origin" in my tests

File: ddos_guard.py [hunk 5]
```diff
@@ -0,0 +1,73 @@
...
+    }
+
+    async with session.get(ddos_guard_url, cookies=cookies, headers=headers_3) as response_3:
+        ddos_guard_cookies = SimpleCookie()
+        ddos_guard_cookies.update(cookies)
+        ddos_guard_cookies.update(response_3.cookies)
```
Comment (e27182) [inline ddos_guard.py]: response_3.cookies contains `__ddg2`, but we have those cookie value when we recieve `check.js`, may be we can skip this call at all?

File: ddoser.py [hunk 6]
```diff
@@ -44,7 +47,11 @@ async def make_request(url: str, proxy: Proxy, timeout: int, headers: Dict[str,
         client_session.headers.update(headers)
 
         async with client_session as session:
-            async with session.get(url, ssl=False, **request_kwargs) as response:
+            cookies = DDOS_GUARD_COOKIE_CACHE.get(base_url)
+            async with session.get(url, cookies=cookies, ssl=False, **request_kwargs) as response:
+                if response.status == HTTPStatus.FORBIDDEN and base_url not in DDOS_GUARD_COOKIE_CACHE:
+                    ddos_guard_cookies = await ddos_guard.bypass(url, response.cookies, session=session)
```
Comment (e27182) [inline ddoser.py]: That way we will call website URL twice:
1. to check for 403
2. in the ddos_guard.bypass
Need to think how to do it just once.

UPDATE: just noticed you reuse cookies, nice!

Comment (e27182) [reply ddoser.py]: In response we will have header `server: ddos-guard`, and then we can just pass response to the ddos_guard.bypass I guess.
No so clean as it is now, but -1 request.

File: ddos_guard.py (hunk 2 above)
Comment (Mystic-Mirage) [reply ddos_guard.py]: The original bypass does just the same. I think ddos-guard validates a site using this cookie. Maybe I'm wrong, can't check it right now.

File: ddos_guard.py (hunk 5 above)
Comment (Mystic-Mirage) [reply ddos_guard.py]: Yes, we can. But if ddos-guard changes the cookie name, it will be broken.

File: ddoser.py (hunk 6 above)
Comment (Mystic-Mirage) [reply ddoser.py]: I think we need to add a check for `server: ddos-guard` because not all 403 errors are ddos-guarded sites.

**Comment by Mystic-Mirage** [comment]
@e27182 Thanks for your efforts!

**Review by e27182 — APPROVED** [review]

**Review by taransergey — APPROVED** [review]

**Pull request merged**